### PR TITLE
8274800: [lworld] Primitive classes can't be retransformed

### PR DESCRIPTION
**Pull request by alexmenkov**
Fix for java.lang.instrument.Instrumentation.retransformClasses()

Cases for retransformation primitive to identity (and identity to primitive) classes and changing primitive class fields (so default_value becomes invalid) are handled by existing code and spec changes are not required.

Also added serviceability test to hotspot_valhalla group

---------
### Progress
- [x] Change must not contain extraneous whitespace

### Issue
 * [JDK-8274800](https://bugs.openjdk.java.net/browse/JDK-8274800): [lworld] Primitive classes can't be retransformed

### Reviewers
 * [Frederic Parain](https://openjdk.java.net/census#fparain) (@fparain - Committer) ⚠️ Review applies to b26c171e6055e21d41ac2561bad902fc376a1c66
 * [Serguei Spitsyn](https://openjdk.java.net/census#sspitsyn) (@sspitsyn - no project role) ⚠️ Review applies to b26c171e6055e21d41ac2561bad902fc376a1c66

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/valhalla pull/588/head:pull/588` \
`$ git checkout pull/588`

Update a local copy of the PR: \
`$ git checkout pull/588` \
`$ git pull https://git.openjdk.java.net/valhalla pull/588/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 588`

View PR using the GUI difftool: \
`$ git pr show -t 588`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/valhalla/pull/588.diff">https://git.openjdk.java.net/valhalla/pull/588.diff</a>

</details>

### CODE REVIEW

**Review by fparain — APPROVED** [review]
Looks good to me.
Thank you for writing the test.

Fred

**Comment by fparain** [comment]
I clicked on approved a bit too quickly.
In jvmtiRedefineClass.cpp, the withfield bytecode has been added, but not the defaultvalue bytecode.

Fred

File: src/hotspot/share/prims/jvmtiClassFileReconstituter.cpp [hunk 1]
```diff
@@ -883,7 +883,7 @@ void JvmtiClassFileReconstituter::write_class_file_format() {
   copy_cpool_bytes(writeable_address(cpool_size()));
 
   // JVMSpec|           u2 access_flags;
-  write_u2(ik()->access_flags().get_flags() & JVM_RECOGNIZED_CLASS_MODIFIERS);
+  write_u2(ik()->access_flags().get_flags() & (JVM_RECOGNIZED_CLASS_MODIFIERS | JVM_ACC_INLINE));
```
Comment (sspitsyn) [inline src/hotspot/share/prims/jvmtiClassFileReconstituter.cpp:886]: Would it better to consider adding JVM_ACC_INLINE to JVM_RECOGNIZED_CLASS_MODIFIERS?
Also, I see other definitions of the JVM_RECOGNIZED_CLASS_MODIFIERS in files:
```
 src/hotspot/share/include/jvm_constants.h
 src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/runtime/ClassConstants.java
```
Do we want this modifier added there as well?

File: test/hotspot/jtreg/TEST.groups [hunk 2]
```diff
@@ -54,7 +54,8 @@ hotspot_serviceability = \
 
 hotspot_valhalla = \
   runtime/valhalla \
-  compiler/valhalla
+  compiler/valhalla \
+  serviceability/jvmti/Valhalla
```
Comment (sspitsyn) [inline test/hotspot/jtreg/TEST.groups:58]: For consistency, should the we rename the test sub-folder from `Valhalla` to `valhalla` ?
This comment is not about this current fix though.

**Review by sspitsyn — APPROVED** [review]

File: src/hotspot/share/prims/jvmtiClassFileReconstituter.cpp (hunk 1 above)
Comment (fparain) [reply src/hotspot/share/prims/jvmtiClassFileReconstituter.cpp:886]: The problem is that JVM_RECOGNIZED_CLASS_MODIFIERS contains all modifiers defined for JDK8 and earlier, but the validity of new modifiers depends on class file version (see classFileParser.cpp where JVM_RECOGNIZED_CLASS_MODIFIERS is also used).

Comment (sspitsyn) [reply src/hotspot/share/prims/jvmtiClassFileReconstituter.cpp:886]: Thanks, Frederic. Of course, it is always possible to add this modifier conditionally depending on the release. But it is not that elegant.

**Comment by alexmenkov** [comment]
> 
> I clicked on approved a bit too quickly. In jvmtiRedefineClass.cpp, the withfield bytecode has been added, but not the defaultvalue bytecode.
> 

I didn't get any failures for defaultvalue. I suppose index in CP is always the same. But I think I need to add it too.
Will try to develop testcase for this

**Comment by alexmenkov** [comment]
> I clicked on approved a bit too quickly. In jvmtiRedefineClass.cpp, the withfield bytecode has been added, but not the defaultvalue bytecode.
> 
Fixed and testcase added.

**Pull request closed without merging**